src/ItemLinkManagement/Public/New-ItemLink.ps1

```diff
@@ -45,7 +45,11 @@ function New-ItemLink {
             "Creating item '$_path'" | Write-Verbose
             New-Item -Path $_path -ItemType $ItemType -Value $_value -Force:$Force
         }catch {
-            Write-Error -Exception $_.Exception -Message $_.Exception.Message -Category $_.CategoryInfo.Category -TargetObject $_.TargetObject
+            if ($ErrorActionPreference -eq 'Stop') {
+                throw
+            }else {
+                Write-Error -ErrorRecord $_
```

Comment:
`Write-Error -ErrorRecord $_` appears to be equivalent to `Write-Error $_` which appears almost equivalent to `Write-Error -Exception $_.Exception -Message $_.Exception.Message -Category $_.CategoryInfo.Category -TargetObject $_.TargetObject` wherein it contains an additional property ` [-ErrorId <string>]`:

```
Write-Error -Exception <Exception> [-Message <string>] [-Category <ErrorCategory>] [-ErrorId <string>] [-TargetObject <Object>] [-RecommendedAction <string>] [-CategoryActivity <string>] [-CategoryReason <string>] [-CategoryTargetName <string>] [-CategoryTargetType <string>] [<CommonParameters>]

Write-Error -ErrorRecord <ErrorRecord> [-RecommendedAction <string>] [-CategoryActivity <string>] [-CategoryReason <string>] [-CategoryTargetName <string>] [-CategoryTargetType <string>] [<CommonParameters>]
```

@leojonathanoh

Reply:
The main thing that is missing is that important steps should return a terminating error on failure which can be caught and rethrown under the function `New-ItemLink.ps1`. In this case, `-ErrorAction Stop` should be added to `New-Item -Path $_path -ItemType $ItemType -Value $_value -Force:$Force`.

```
$ Import-Module .\src\ItemLinkManagement\ItemLinkManagement.psm1 -Force -Verbose
$ 'abc','abcd' | % { New-ItemLink -Path $_ -ItemType Junction -value "$(get-item .\src)" }

New-ItemLink : The file or directory is not a reparse point
At path\to\theohbrothers\ItemLinkManagement\debugexception.ps1:4 char:20
+ ... abcd' | % { New-ItemLink -Path $_ -ItemType Junction -value "$(get-it ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ CategoryInfo          : WriteError: (D:\git\theohbrother\u2026mLinkManagement\abc:String) [Write-Error], Win32Exception
+ FullyQualifiedErrorId : System.ComponentModel.Win32Exception,New-ItemLink
```

Reply:
Rather `$ErrorActionPreference = 'Stop'` should be specified within the function with `-ErrorAction SilentlyContinue` used in strategic locations: 

* https://github.com/startersclan/Compile-SourceScript/blob/9b8d565cca83c10f9b127b9f91203e51f60d8cd8/src/Compile-SourceScript/Public/Compile-SourceScript.ps1#L45
* https://github.com/startersclan/Compile-SourceScript/blob/9b8d565cca83c10f9b127b9f91203e51f60d8cd8/src/Compile-SourceScript/Public/Compile-SourceScript.ps1#L115

Reply:
The two tests under `Context 'Error stream'` are failing now. Unsure what is it about the exceptions that the tests are failing because error records are returned from each scenario: https://dev.azure.com/theohbrothers/ItemLinkManagement/_build/results?buildId=341&view=logs&j=cf913be4-6704-58f5-cde6-258cc8aef7bf&t=c3dd6012-84fa-5b88-85db-ffd0f16e84df&l=519

@leojonathanoh

Reply:
lol I thought these were posted for you to see. Had to confirm review before the comments are sent

Reply:
@joeltimothyoh well u should know what to change in the unit tests which are important to ensuring expected function behavior remains the same even with slightest change.

Reply:
> The two tests under Context 'Error stream' are failing now. Unsure what is it about the exceptions that the tests are failing because error records are returned from each scenario: https://dev.azure.com/theohbrothers/ItemLinkManagement/_build/results?buildId=341&view=logs&j=cf913be4-6704-58f5-cde6-258cc8aef7bf&t=c3dd6012-84fa-5b88-85db-ffd0f16e84df&l=519

It's because the errors have become terminating.

Reply:
> It's because the errors have become terminating.

Nope. They are returned as non-terminating by `Write-Error`. @leojonathanoh

Reply:
@joeltimothyoh they are terminating, because of `$ErrorActionPreference = 'Stop'` occurring at the beginning of the function.

Reply:
btw

```
$ErrorActionPreference = 'Stop'
Write-Error 'foo'

# is the same as
Write-Error 'foo' -ErrorAction 'Stop'
```

both result in a terminating error, and the stack trace shows the `Write-Error` line itself.

Reply:
> @joeltimothyoh they are terminating, because of $ErrorActionPreference = 'Stop' occurring at the beginning of the function.

They are terminating inside the try-catch block in `New-ItemLink`, but returned as non-terminating by the function with `Write-Error`. @leojonathanoh

Reply:
> They are terminating inside the try-catch block in New-ItemLink, but returned as non-terminating by the function with Write-Error. @leojonathanoh

Hmm for some reason the error object returned is in fact terminating. Trying to figure why because it shouldn't be from my experience. @leojonathanoh

Reply:
I figured.

Quoting from the [link](https://powershellexplained.com/2017-04-10-Powershell-exceptions-everything-you-ever-wanted-to-know/#exception) you sent:

> I mentioned that Write-Error does not throw a terminating error by default. If you specify -ErrorAction Stop then Write-Errorgenerates a terminating error that can be handled with a catch.

So you're right that the tests are failing now because the errors returned are terminating. So all that needs to be done is to make the function default to terminating but specify `-ErrorAction Continue` with `Write-Error`. @leojonathanoh 

```